nvme.c

```diff
@@ -221,67 +217,53 @@ void *nvme_realloc(void *p, size_t len)
 	return result;
 }
 
-static void *__nvme_alloc_huge(size_t len, bool *huge)
-{
-	void *p;
-
-	if (!posix_memalign(&p, getpagesize(), len)) {
-		*huge = false;
-		memset(p, 0, len);
-		return p;
-	}
-	return NULL;
-}
-
 #define HUGE_MIN 0x80000
 
-#ifdef CONFIG_LIBHUGETLBFS
-void nvme_free_huge(void *p, bool huge)
+void nvme_free_huge(void *p, size_t len)
 {
-	if (huge) {
-		if (p)
-			free_hugepage_region(p);
-	} else {
+	size_t _len = ROUND_UP(len, 0x1000);
+
+	if (_len < HUGE_MIN) {
 		free(p);
+		return;
 	}
+
+	munmap(p, _len);
 }
 
-void *nvme_alloc_huge(size_t len, bool *huge)
+void *nvme_alloc_huge(size_t len)
 {
+	size_t _len = ROUND_UP(len, 0x1000);
 	void *p;
 
-	if (len < HUGE_MIN)
-		return __nvme_alloc_huge(len, huge);
+	if (_len < HUGE_MIN) {
+		if (!posix_memalign(&p, getpagesize(), len)) {
+			memset(p, 0, len);
+			return p;
+		}
+	}
 
-	p = get_hugepage_region(len, GHR_DEFAULT);
-	if (!p)
-		return __nvme_alloc_huge(len, huge);
+	p = mmap(NULL, _len, PROT_READ | PROT_WRITE,
+		 MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+	if (p < 0)
```

Comment:
Don't think it is possible for a pointer to be negative? Should this be `if (p == MAP_FAILED)`?

Reply:
Need to also include '`MAP_HUGETLB`' right?

Reply:
- `MAP_FAILED` is according the man pages `-1` but yes looks better with the define.
- `MAP_HUGETLB`: was not sure if we the flag and `madvise` at the same time. From the man pages I'd get the impression either the mmap flag or the `madvise` call should do the same thing. But I am really just guessing.

Reply:
> And I found following example: https://github.com/torvalds/linux/blob/master/tools/testing/selftests/mm/hugepage-mmap.c
> 
> So I'd say it's okay just to follow the kernel example and don't implement any of the fs-based fallback strategy.

I agree, fs-based fallbacks need not be considered here.

The kernel test example you pointed to using memfd_create()'s flags appear to have the same effects as doing an anonymous mmap. I considered that way since I'm using memfd's extensively in another project, but I don't see the point if we're don't share it with other processes. Unless I'm missing something, I think memfd's are just an unnecessary extra step for nvme-cli's usage.